tests/test_quaternion.h

```diff
@@ -0,0 +1,366 @@
+	double roll = deg2rad(10.0);
+
+	Vector3 euler_y(0.0, yaw, 0.0);
+	Quaternion q_y(euler_y);
+	CHECK(q_y[0] == doctest::Approx(0.0));
+	CHECK(q_y[1] == doctest::Approx(0.382684));
+	CHECK(q_y[2] == doctest::Approx(0.0));
+	CHECK(q_y[3] == doctest::Approx(0.923879));
+
+	Vector3 euler_p(pitch, 0.0, 0.0);
+	Quaternion q_p(euler_p);
+	CHECK(q_p[0] == doctest::Approx(0.258819));
+	CHECK(q_p[1] == doctest::Approx(0.0));
+	CHECK(q_p[2] == doctest::Approx(0.0));
+	CHECK(q_p[3] == doctest::Approx(0.965926));
+
+	Vector3 euler_r(0.0, 0.0, roll);
+	Quaternion q_r(euler_r);
+	CHECK(q_r[0] == doctest::Approx(0.0));
+	CHECK(q_r[1] == doctest::Approx(0.0));
+	CHECK(q_r[2] == doctest::Approx(0.0871558));
+	CHECK(q_r[3] == doctest::Approx(0.996195));
+}
+
+TEST_CASE("[Quaternion] Construct Euler YXZ dynamic axes") {
+	double yaw = deg2rad(45.0);
+	double pitch = deg2rad(30.0);
+	double roll = deg2rad(10.0);
+
+	// Generate YXZ comparision data (Z-then-X-then-Y) using
+	//   single-axis Euler constructor and quaternion product,
+	//   both tested separately.
+	Vector3 euler_y(0.0, yaw, 0.0);
+	Quaternion q_y(euler_y);
+	Vector3 euler_p(pitch, 0.0, 0.0);
+	Quaternion q_p(euler_p);
+	Vector3 euler_r(0.0, 0.0, roll);
+	Quaternion q_r(euler_r);
+
+	// Roll-Z is followed by Pitch-X.
+	Quaternion check_xz = q_p * q_r;
+	// Then Yaw-Y follows both.
+	Quaternion check_yxz = q_y * check_xz;
+
+	// Test construction from YXZ Euler angles.
+	Vector3 euler_yxz(pitch, yaw, roll);
+	Quaternion q(euler_yxz);
+	CHECK(q[0] == doctest::Approx(check_yxz[0]));
+	CHECK(q[1] == doctest::Approx(check_yxz[1]));
+	CHECK(q[2] == doctest::Approx(check_yxz[2]));
+	CHECK(q[3] == doctest::Approx(check_yxz[3]));
+
+	// Sneak in a test of is_equal_approx.
+	CHECK(q.is_equal_approx(check_yxz));
+}
+
+TEST_CASE("[Quaternion] Construct Basis Euler") {
+	double yaw = deg2rad(45.0);
+	double pitch = deg2rad(30.0);
+	double roll = deg2rad(10.0);
+	Vector3 euler_yxz(pitch, yaw, roll);
+	Quaternion q_yxz(euler_yxz);
+	//
+	Basis basis_axes(euler_yxz);
+	Quaternion q(basis_axes);
+	CHECK(q.is_equal_approx(q_yxz));
+}
+
+TEST_CASE("[Quaternion] Construct Basis Axes") {
+	// Arbitrary Euler angles.
+	Vector3 euler_yxz(deg2rad(31.41), deg2rad(-49.16), deg2rad(12.34));
+	// Basis vectors from online calculation of rotation matrix.
+	Vector3 i_unit(0.5545787, 0.1823950, 0.8118957);
+	Vector3 j_unit(-0.5249245, 0.8337420, 0.1712555);
+	Vector3 k_unit(-0.6456754, -0.5211586, 0.5581192);
+	// Quaternion from online calculation.
+	Quaternion q_calc(0.2016913, -0.4245716, 0.206033, 0.8582598);
+	// Quaternion from local calculation.
+	Quaternion q_local = quat_euler_yxz_deg(Vector3(31.41, -49.16, 12.34));
+	// Quaternion from Euler angles constructor.
+	Quaternion q_euler(euler_yxz);
+	CHECK(q_calc.is_equal_approx(q_local));
+	CHECK(q_local.is_equal_approx(q_euler));
+
+	// Calculate Basis and construct Quaternion
+	// TODO: Restore the Basis construction from axes when bug is fixed,
```

Comment:
Is this todo done?

Reply:
The TODO was based on my misunderstanding of Basis, so was OBE. I reworded the comment to correct it. Thanks for noticing.